### PR TITLE
Feature/refactor search snapshots

### PR DESCRIPTION
**Pull request by gnusi**
### Scope & Purpose

Refactor management of ArangoSearch index snapshot. This PR greatly reduces load to IO eliminating unnecessary  syscalls. (`mmap`/`fstat`/etc).

- [ ] :hankey: Bugfix
- [ ] :pizza: New feature
- [x] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.10: *(Please link PR)*
  - [ ] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [x] Enterprise PR: https://github.com/arangodb/enterprise/pull/1193
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

**Review by Dronplane — APPROVED** [review]
LGTM

File: arangod/IResearch/IResearchDataStore.cpp [hunk 1]
```diff
@@ -1415,51 +1377,57 @@ Result IResearchDataStore::initDataStore(
...
-    _dataStore._writer->commit();  // initialize 'store'
-    _dataStore._reader = irs::directory_reader::open(*(_dataStore._directory),
-                                                     _format, readerOptions);
+  if (!pathExists) {
+    _dataStore._writer->Commit();  // Initialize 'store'
   }
 
+  _dataStore._reader = _dataStore._writer->GetSnapshot();
+
   if (!_dataStore._reader) {
```
Comment (MBkkt) [inline arangod/IResearch/IResearchDataStore.cpp:1401]: Can it happens or it is assert?

Comment (MBkkt) [inline arangod/IResearch/IResearchDataStore.cpp:1401]: Can it happens or it is assert?

**Review by MBkkt — CHANGES_REQUESTED** [review]

File: CMakeLists.txt [hunk 2]
```diff
@@ -686,7 +686,8 @@ option(USE_MAINTAINER_MODE
 )
 
 if (USE_MAINTAINER_MODE)
-  set(IRESEARCH_DEBUG ON)
+  add_definitions("-DIRESEARCH_DEBUG")
```
Comment (jsteemann) [inline CMakeLists.txt:689]: Nit:
By setting the definition here, we set it globally, so it will leak into other modules (e.g. RocksDB). Although in other modules it will not do any harm, because the define name has an `IRESEARCH_` prefix.

Potential improvement:
I guess this definition mostly affects code in the `3rdParty/iresearch` subdirectory. But as the define is picked up in headers files in `3rdParty/iresearch` that we include from the ArangoDB code, we need it in ArangoDB too. 
We could set it for specific targets (e.g. some targets inside ArangoDB and iresearch) only.

But it should already work as it is now.

File: tests/IResearch/IResearchQueryNGramMatchTest.cpp [hunk 3]
```diff
@@ -35,7 +35,6 @@
 #include "VocBase/LogicalCollection.h"
 #include "store/mmap_directory.hpp"
 #include "utils/index_utils.hpp"
-#include "utils/string_utils.hpp"
```
Comment (jsteemann) [inline tests/IResearch/IResearchQueryNGramMatchTest.cpp:38]: absl::Substitute is now used by this test. Should the corresponding header file be included here?

File: tests/IResearch/IResearchQueryLevenshteinMatchTest.cpp [hunk 4]
```diff
@@ -35,7 +35,6 @@
 #include "VocBase/LogicalCollection.h"
 #include "store/mmap_directory.hpp"
 #include "utils/index_utils.hpp"
-#include "utils/string_utils.hpp"
```
Comment (jsteemann) [inline tests/IResearch/IResearchQueryLevenshteinMatchTest.cpp:38]: absl::Substitute is now used by this test. Should the corresponding header file be included here?

File: tests/IResearch/IResearchQueryOrTest.cpp [hunk 5]
```diff
@@ -36,7 +36,6 @@
 #include "VocBase/LogicalCollection.h"
 #include "store/mmap_directory.hpp"
 #include "utils/index_utils.hpp"
-#include "utils/string_utils.hpp"
```
Comment (jsteemann) [inline tests/IResearch/IResearchQueryOrTest.cpp:39]: absl::Substitute is now used by this test. Should the corresponding header file be included here?

File: tests/IResearch/IResearchQueryPhraseTest.cpp [hunk 6]
```diff
@@ -37,7 +37,6 @@
 #include "common.h"
 #include "store/mmap_directory.hpp"
 #include "utils/index_utils.hpp"
-#include "utils/string_utils.hpp"
```
Comment (jsteemann) [inline tests/IResearch/IResearchQueryPhraseTest.cpp:40]: absl::Substitute is now used by this test. Should the corresponding header file be included here?

File: tests/IResearch/IResearchQuerySelectAllTest.cpp [hunk 7]
```diff
@@ -36,7 +36,6 @@
 #include "VocBase/LogicalCollection.h"
 #include "store/mmap_directory.hpp"
 #include "utils/index_utils.hpp"
-#include "utils/string_utils.hpp"
```
Comment (jsteemann) [inline tests/IResearch/IResearchQuerySelectAllTest.cpp:39]: absl::Substitute is now used by this test. Should the corresponding header file be included here?

File: tests/IResearch/IResearchQueryStartsWithTest.cpp [hunk 8]
```diff
@@ -36,7 +36,6 @@
 #include "VocBase/LogicalCollection.h"
 #include "store/mmap_directory.hpp"
 #include "utils/index_utils.hpp"
-#include "utils/string_utils.hpp"
```
Comment (jsteemann) [inline tests/IResearch/IResearchQueryStartsWithTest.cpp:39]: absl::Substitute is now used by this test. Should the corresponding header file be included here?

File: arangod/IResearch/IResearchDataStore.cpp [hunk 9]
```diff
@@ -1415,51 +1377,57 @@ Result IResearchDataStore::initDataStore(
...
+            absl::StrCat("Failed to get last committed tick while initializing "
                          "ArangoSearch index '",
                          index().id().id(), "'")};
   }
 
-  LOG_TOPIC("7e128", TRACE, TOPIC)
-      << "data store reader for link '" << index().id()
-      << "' is initialized with recovery tick low '"
+  LOG_TOPIC("7e028", TRACE, TOPIC)
+      << "Successfully opened existing data store data store reader for "
```
Comment (jsteemann) [inline arangod/IResearch/IResearchDataStore.cpp:1420]: Is the log message now still appropriate? It says "opened existing data store", but I think we get here even if no data store existed before? Should the message depend on `pathExists`, e.g.
```suggestion
if (pathExists) {
  LOG_TOPIC(...) << "Successfully opened existing data store ..."
} else {
  LOG_TOPIC(...) << "Successfully initialized new data store ..."
}
```

Comment (jsteemann) [inline arangod/IResearch/IResearchDataStore.cpp:1420]: The current message also reads `data store data store reader`, which is hard to parse. What about just `data store reader`?

File: arangod/IResearch/IResearchDataStore.h [hunk 10]
```diff
@@ -301,7 +300,7 @@ class IResearchDataStore {
 
     void resetDataStore() noexcept {
       // reset all underlying readers to release file handles
-      _reader.reset();
+      _reader = {};
```
[the remark this answers is not in the document]
Comment (gnusi) [reply arangod/IResearch/IResearchDataStore.h:303]: Yes, I did.

File: arangod/IResearch/IResearchDataStore.cpp [hunk 11]
```diff
@@ -1415,51 +1377,57 @@ Result IResearchDataStore::initDataStore(
...
 
-  if (!_dataStore._reader) {
-    _dataStore._writer->commit();  // initialize 'store'
-    _dataStore._reader = irs::directory_reader::open(*(_dataStore._directory),
-                                                     _format, readerOptions);
+  if (!pathExists) {
+    _dataStore._writer->Commit();  // Initialize 'store'
   }
 
+  _dataStore._reader = _dataStore._writer->GetSnapshot();
```
[the remark this answers is not in the document]
Comment (gnusi) [reply arangod/IResearch/IResearchDataStore.cpp:1399]: It's always valid, but there is a commit above in this case.

File: arangod/IResearch/IResearchDataStore.cpp (hunk 1 above)
Comment (gnusi) [reply arangod/IResearch/IResearchDataStore.cpp:1401]: Yes, can be turned into an assertion.

File: arangod/IResearch/IResearchDataStore.cpp [hunk 12]
```diff
@@ -1415,51 +1377,57 @@ Result IResearchDataStore::initDataStore(
...
   }
 
-  LOG_TOPIC("7e128", TRACE, TOPIC)
-      << "data store reader for link '" << index().id()
-      << "' is initialized with recovery tick low '"
+  LOG_TOPIC("7e028", TRACE, TOPIC)
+      << "Successfully opened existing data store data store reader for "
+      << "ArangoSearch index '" << index().id() << "', docs count '"
+      << _dataStore._reader->docs_count() << "', live docs count '"
+      << _dataStore._reader->live_docs_count() << "', recovery tick low '"
```
[the remark this answers is not in the document]
Comment (gnusi) [reply arangod/IResearch/IResearchDataStore.cpp:1423]: Let's address it separately

File: tests/IResearch/IResearchQueryNGramMatchTest.cpp (hunk 3 above)
Comment (gnusi) [reply tests/IResearch/IResearchQueryNGramMatchTest.cpp:38]: :heavy_check_mark:

File: arangod/IResearch/IResearchDataStore.cpp (hunk 9 above)
Comment (gnusi) [reply arangod/IResearch/IResearchDataStore.cpp]: :heavy_check_mark:

Comment (gnusi) [reply arangod/IResearch/IResearchDataStore.cpp]: Done

File: CMakeLists.txt (hunk 2 above)
Comment (gnusi) [reply CMakeLists.txt:689]: I made it global because some Arango headers include code relying on `IRESEARCH_DEBUG` macro and there was a very subtle issue related to that.

Comment (MBkkt) [reply CMakeLists.txt:689]: @jsteemann I think we can improve it by write in the iresearch submodule

```cmake
target_compile_definitions(some_base_iresearch_target PUBLIC -DIRESEARCH_DEBUG)
```

Comment (MBkkt) [reply CMakeLists.txt:689]: But from the other side, if someone don't link against iresearch and use header only files from iresearch it will be broken

File: arangod/IResearch/IResearchDataStore.cpp [hunk 13]
```diff
@@ -1051,7 +1065,7 @@ Result IResearchDataStore::commitUnsafeImpl(
 
     // update reader
     TRI_ASSERT(_dataStore._reader != reader);
-    _dataStore._reader = reader;
+    _dataStore._reader = std::move(reader);
```
Comment (MBkkt) [inline arangod/IResearch/IResearchDataStore.cpp:1068]:
```suggestion
    _dataStore._reader = reader;
```

**Review by MBkkt — CHANGES_REQUESTED** [review]

**Review by MBkkt — APPROVED** [review]
